src/core/thread/mle.cpp

```diff
@@ -5060,30 +5049,31 @@ Error Mle::TxMessage::AppendAddressRegistrationTlv(AddressRegistrationMode aMode
     return error;
 }
 
-Error Mle::TxMessage::AppendCompressedAddressEntry(uint8_t aContextId, const Ip6::Address &aAddress)
+Error Mle::TxMessage::AppendAddressRegistrationEntry(const Ip6::Address &aAddress)
 {
-    // Append an IPv6 address entry in an Address Registration TLV
-    // using compressed format (context ID with IID).
-
-    Error error;
-
-    SuccessOrExit(error = Append<uint8_t>(AddressRegistrationTlv::ControlByteFor(aContextId)));
-    error = Append(aAddress.GetIid());
+    uint8_t ctlByte = AddressRegistrationTlv::kControlByteUncompressed;
+    Error   error;
 
-exit:
-    return error;
-}
+    if (!aAddress.IsMulticast())
+    {
+        Lowpan::Context context;
 
-Error Mle::TxMessage::AppendAddressEntry(const Ip6::Address &aAddress)
-{
-    // Append an IPv6 address entry in an Address Registration TLV
-    // using uncompressed format
+        if ((Get<NetworkData::Leader>().GetContext(aAddress, context) == kErrorNone) && context.mCompressFlag)
+        {
+            ctlByte = AddressRegistrationTlv::ControlByteFor(context.mContextId);
+        }
+    }
 
-    Error   error;
-    uint8_t controlByte = AddressRegistrationTlv::kControlByteUncompressed;
+    SuccessOrExit(error = Append(ctlByte));
 
-    SuccessOrExit(error = Append(controlByte));
-    error = Append(aAddress);
+    if (ctlByte == AddressRegistrationTlv::kControlByteUncompressed)
```

Comment:
Need to check specifically the compressed flag, right?

Reply:
Yes. We check it above before setting the `ctlByte`. 
```c++
        if ((Get<NetworkData::Leader>().GetContext(aAddress, context) == kErrorNone) && context.mCompressFlag)
        {
            ctlByte = AddressRegistrationTlv::ControlByteFor(context.mContextId);
        }
    }
```